compiler/backend/src/org/jetbrains/kotlin/codegen/inline/MethodInliner.kt

```diff
@@ -305,51 +309,33 @@ class MethodInliner(
                     inlineOnlySmapSkipper?.markCallSiteLineNumber(remappingMethodAdapter)
                 } else if (isAnonymousConstructorCall(owner, name)) { //TODO add method
                     //TODO add proper message
-                    assert(transformationInfo is AnonymousObjectTransformationInfo) {
+                    var info = transformationInfo as? AnonymousObjectTransformationInfo ?: throw AssertionError(
                         "<init> call doesn't correspond to object transformation info for '$owner.$name': $transformationInfo"
+                    )
+                    val ownerIsParentClass = inlineCallSiteInfo.ownerClassName == info.oldClassName
+                    if (ownerIsParentClass) {
+                        // inline fun f() -> new f$1 -> fun something() in class f$1 -> new f$1
+                        //                   ^-- fetch the info that was created for this instruction
+                        info = inliningContext.parent?.transformationInfo as? AnonymousObjectTransformationInfo
```

Comment:
@pyos As you have fresh look: could you point to  logic where compiler checks that object is already in process of regeneration and don't try regenerate it on second turn?

Reply:
In `handleAnonymousObjectRegeneration`, `remapper.addMapping` is done before invoking the transformer. Thus, the nested `MethodInliner`'s `isAlreadyRegenerated` will return `true` for this type, causing `handleAnonymousObjectRegeneration` to ignore it. I think that's the only case where `isAlreadyRegenerated` returns true, actually...